src/_pytest/cacheprovider.py

```diff
@@ -203,17 +203,32 @@ def set(self, key: str, value: object) -> None:
             with f:
                 f.write(data)
 
-    def _ensure_supporting_files(self) -> None:
-        """Create supporting files in the cache dir that are not really part of the cache."""
-        readme_path = self._cachedir / "README.md"
-        readme_path.write_text(README_CONTENT, encoding="UTF-8")
-
-        gitignore_path = self._cachedir.joinpath(".gitignore")
-        msg = "# Created by pytest automatically.\n*\n"
-        gitignore_path.write_text(msg, encoding="UTF-8")
+    def _ensure_cache_dir_and_supporting_files(self) -> None:
+        """Create the cache dir and its supporting files."""
+        if self._cachedir.is_dir():
+            return
 
-        cachedir_tag_path = self._cachedir.joinpath("CACHEDIR.TAG")
-        cachedir_tag_path.write_bytes(CACHEDIR_TAG_CONTENT)
+        with tempfile.TemporaryDirectory() as d:
+            path = Path(d)
+            with open(path.joinpath("README.md"), "xt", encoding="UTF-8") as f:
+                f.write(README_CONTENT)
+            with open(path.joinpath(".gitignore"), "xt", encoding="UTF-8") as f:
+                f.write("# Created by pytest automatically.\n*\n")
+            with open(path.joinpath("CACHEDIR.TAG"), "xb") as f:
+                f.write(CACHEDIR_TAG_CONTENT)
+
+            self._cachedir.parent.mkdir(parents=True, exist_ok=True)
+            try:
+                os.rename(d, self._cachedir)
+                # Create a directory in place of the one we just moved so that
+                # `TemporaryDirectory`'s cleanup doesn't complain.
+                #
+                # TODO: pass ignore_cleanup_errors=True when we no longer support python < 3.10. See
+                # https://github.com/python/cpython/issues/74168. Note that passing delete=False
+                # would do the wrong thing in case of errors and isn't supported until python 3.12.
+                os.mkdir(d)
```

Comment:
```suggestion
                d.mkdir()
```

It would be good to define how we deal with situation when directory already exists.

Reply:
We just renamed it on the previous line, so it shouldn't exist. If it does, we should probably allow the exception to bubble up. Done.